File: apps/node/src/scraper.ts
import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { z, ZodSchema } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';
import dotenv from 'dotenv'
import path from 'path'

// Configure dotenv to look for .env file in the root directory
dotenv.config({ path: path.resolve(__dirname, '../../../.env') })

type ReasoningEffort = 'medium' | 'high';

export class ScraperClient {
  private baseUrl: string;
  private client: AxiosInstance;

  constructor(baseUrl: string = `${process.env.BASE_API_URL}:${process.env.BASE_API_PORT}`) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.client = axios.create({
      baseURL: this.baseUrl,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  /**
   * Scrape a webpage and return a validated result
   */
  async scrape<T>(
    url: string, 
    schema: ZodSchema<T>, 
    timeout: number = Number(process.env.DEFAULT_TIMEOUT),
    reasoning_effort?: ReasoningEffort,
    prompt?: string,
    top_p?: number,
    temperature?: number
  ): Promise<T> {
    try {
      // Convert Zod schema to JSON schema using the library
      const jsonSchema = zodToJsonSchema(schema);
      
      const payload: any = {
        url,
        schema: jsonSchema,
      };

      // Add optional parameters to payload if provided
      if (reasoning_effort !== undefined) {
        payload.reasoning_effort = reasoning_effort;
      }
      if (prompt !== undefined) {
        payload.prompt = prompt;
      }
      if (top_p !== undefined) {
        payload.top_p = top_p;
      }
      if (temperature !== undefined) {
        payload.temperature = temperature;
      }

      const response: AxiosResponse = await this.client.post('/scrape', payload, {
        timeout,
      });

      const result = response.data;
      
      if (!result.success) {
        throw new Error(`API returned error: ${result.error || 'Unknown error'}`);
      }

      // Validate and return the result using Zod
      return schema.parse(result.data);
      
    } catch (error) {
      if (error instanceof z.ZodError) {
        throw new Error(`Validation failed: ${error.message}`);
      }
      if (axios.isAxiosError(error)) {
        throw new Error(`API request failed: ${error.message}`);
      }
      throw error;
    }
  }
}

File: apps/python/src/scraper.py
import requests
from typing import Dict, Any, Type, Optional, Literal, TypeVar, Union
import json
from pydantic import BaseModel
import os
from dotenv import load_dotenv

load_dotenv()

# Define a TypeVar bound to BaseModel
T = TypeVar('T', bound=BaseModel)

class BitBuffet:
    """Python SDK for the Structured Scraper API"""
    
    def __init__(self, api_key: str):
        # Validate API key
        if not api_key or not api_key.strip():
            raise ValueError('API key is required. Please provide a valid API key when initializing the BitBuffet.')
        
        self.base_url = f"{os.getenv('BASE_API_URL')}:{os.getenv('BASE_API_PORT')}"
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json'
        })
    
    def _pydantic_to_json_schema(self, schema_class: Type[BaseModel]) -> Dict[str, Any]:
        """Convert a Pydantic model to JSON schema format"""
        schema = schema_class.model_json_schema()
        
        # Resolve $ref references to inline definitions
        if '$defs' in schema:
            def resolve_refs(obj, defs):
                if isinstance(obj, dict):
                    if '$ref' in obj:
                        ref_path = obj['$ref'].split('/')[-1]
                        return resolve_refs(defs[ref_path], defs)
                    else:
                        return {k: resolve_refs(v, defs) for k, v in obj.items()}
                elif isinstance(obj, list):
                    return [resolve_refs(item, defs) for item in obj]
                return obj
            
            schema = resolve_refs(schema, schema['$defs'])
            schema.pop('$defs', None)
        
        return schema
    
    def scrape(
        self, 
        url: str, 
        schema_class: Type[T], 
        timeout: int = int(os.getenv('DEFAULT_TIMEOUT'))/1000,
        reasoning_effort: Optional[Literal['medium', 'high']] = None,
        prompt: Optional[str] = None,
        top_p: Optional[Union[int, float]] = None,
        temperature: Optional[Union[int, float]] = None
    ) -> T:
        """
        Scrape a webpage and return a validated Pydantic model instance.
        
        Args:
            url: The URL to scrape
            schema_class: Pydantic BaseModel class to validate against
            timeout: Request timeout in seconds
            reasoning_effort: Reasoning effort level ('medium' or 'high')
            prompt: Additional custom prompt for extra configurability
            top_p: Top-p sampling parameter for the AI model
            temperature: Temperature parameter for the AI model (0-1.5)
            
        Returns:
            Validated Pydantic model instance of the same type as schema_class
            
        Raises:
            requests.RequestException: If the API request fails
            ValueError: If the response is invalid or both temperature and top_p are provided
        """
        # Validate that both temperature and top_p are not provided simultaneously
        if temperature is not None and top_p is not None:
            raise ValueError("Cannot specify both 'temperature' and 'top_p' parameters. Please use only one.")
        
        # Convert Pydantic schema to JSON schema
        schema = self._pydantic_to_json_schema(schema_class)
        
        payload = {
            "url": url,
            "json_schema": schema
        }
        
        # Add optional parameters to payload if provided
        if reasoning_effort is not None:
            payload["reasoning_effort"] = reasoning_effort
        if prompt is not None:
            payload["prompt"] = prompt
        if top_p is not None:
            payload["top_p"] = top_p
        if temperature is not None:
            payload["temperature"] = temperature
        
        try:
            response = self.session.post(
                f"{self.base_url}/scrape",
                json=payload,
                timeout=timeout,
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            
            try:
                result = response.json()
            except (json.JSONDecodeError, ValueError) as e:
                raise ValueError(f"Invalid JSON response: {e}")
                
            if not result.get('success'):
                raise ValueError(f"API returned error: {result.get('error', 'Unknown error')}")
            
            # Return validated Pydantic model instance
            return schema_class.model_validate(result['data'])
            
        except requests.RequestException as e:
            raise requests.RequestException(f"API request failed: {e}")